scripts/gen_dockerfile.py

```diff
@@ -0,0 +1,250 @@
+PYTHON_INTERPRETER_VERSION_MAP = {
+    '': '',  # == 2.7
+    '2': '',  # == 2.7
+    '3': '3.5',
+    '3.4': '3.4',
+    '3.5': '3.5',
+    '3.6': '3.6',
+}
+
+# File templates.
+# Designed to exactly match the current output of 'gcloud app gen-config'
+DOCKERFILE_TEMPLATE = """\
+FROM {base_image}
+LABEL python_version=python{dockerfile_python_version}
+RUN virtualenv --no-download /env -p python{dockerfile_python_version}
+
+# Set virtualenv environment variables. This is equivalent to running
+# source /env/bin/activate
+ENV VIRTUAL_ENV /env
+ENV PATH /env/bin:$PATH
+{optional_requirements_txt}ADD . /app/
+{optional_entrypoint}"""
+
+DOCKERFILE_REQUIREMENTS_TXT = """\
+ADD requirements.txt /app/
+RUN pip install -r requirements.txt
+"""
+
+DOCKERFILE_ENTRYPOINT_TEMPLATE = """\
+CMD {entrypoint}
+"""
+
+DOCKERIGNORE = """\
+# Copyright 2015 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+.dockerignore
+Dockerfile
+.git
+.hg
+.svn
+"""
+
+# Validated application configuration
+AppConfig = collections.namedtuple(
+    'AppConfig',
+    'base_image dockerfile_python_version entrypoint has_requirements_txt'
+)
+
+
+def get_app_config(raw_config, args):
+    """Read and validate the application runtime configuration.
+
+    Args:
+        raw_config (dict): deserialized app.yaml
+        args (argparse.Namespace): command line flags
+
+    Returns:
+        AppConfig: valid configuration
+    """
+    # Examine app.yaml
+    if not isinstance(raw_config, dict):
+        raise ValueError(
+            'Expected {} contents to be of type "dict", but found type "{}"'.
+            format(args.config, type(raw_config)))
+
+    entrypoint = validation_utils.get_field_value(raw_config, 'entrypoint', str)
+    if not PRINTABLE_REGEX.match(entrypoint):
+        raise ValueError('Invalid character in "entrypoint" field of app.yaml')
+    raw_runtime_config = validation_utils.get_field_value(raw_config, 'runtime_config', dict)
+    python_version = validation_utils.get_field_value(raw_runtime_config, 'python_version', str)
+    dockerfile_python_version = PYTHON_INTERPRETER_VERSION_MAP.get(
+        python_version)
+    if dockerfile_python_version is None:
+        valid_versions = str(sorted(PYTHON_INTERPRETER_VERSION_MAP.keys()))
+        msg = ('Invalid "python_version" field in "runtime_config" section '
+               'of app.yaml.  Valid options are:\n{}').format(valid_versions)
+        raise ValueError(msg)
```

Comment:
silly nit: I prefer to just do continuations inside of the `ValueError(` instead of a temporary variable, but no action needed here.